src/content/developers/tutorials/merkle-proofs-for-offline-data-integrity/index.md

```diff
@@ -0,0 +1,253 @@
+
+```solidity
+
+contract MerkleProof {
+    uint merkleRoot;
+
+    function getRoot() public view returns (uint) {
+      return merkleRoot;
+    }
+
+    // Extremely insecure, in production code access to
+    // this function MUST BE strictly limited, probably to an
+    // owner
+    function setRoot(uint _merkleRoot) external {
+      merkleRoot = _merkleRoot;
+    }   // setRoot
+```
+
+Set and get functions for the Merkle root. Letting everybody update the Merkle root is an *extremely bad idea* in a production system. I do it here for the sake of simplicity for sample code. **Don't do it on a system where data integrity actually matters**.
+
+```solidity
+    function pairHash(uint _a, uint _b) internal pure returns(uint) {
+      return uint(keccak256(abi.encode(_a ^ _b)));
+    }
+```
+
+This function generates a pair hash. It is just the Solidity transalation of the JavaScript code for `pairHash`. 
+
+**Note:** This is another case of optimization for readability. Based on [the function definition](https://www.tutorialspoint.com/solidity/solidity_cryptographic_functions.htm), it might be possible to store the data as a [`bytes32`](https://docs.soliditylang.org/en/v0.5.3/types.html#fixed-size-byte-arrays) value and avoid the conversions. 
+
+```solidity
+    // Verify a Merkle proof
+    function verifyProof(uint _value, uint[] calldata _proof) 
+        public view returns (bool) {
+      uint temp = _value;
+      uint i;
+  
+      for(i=0; i<_proof.length; i++) {
+        temp = pairHash(temp, _proof[i]);
+      }
+
+      return temp == merkleRoot;
+    }
+    
+}  // MarkleProof
+```
+
+In mathematical notation Merkle proof verification looks like this: `H(proof_n, H(proof_n-1, H(proof_n-2, ... H(proof_1, H(proof_0, value))...)))`. This code implements it.
+
+
+
+## Merkle proofs and rollups don't mix
+
+Merkle proofs don't work well with [rollups](https://ethereum.org/en/developers/docs/scaling/layer-2-rollups/). The reason is that rollups write all the transaction data on L1, but process on L2. The cost to send a Merkle proof with a transaction averages to 638 gas per layer (currently a byte in call data costs 16 gas if it isn't zero, and 4 if it is zero). If we have 1024 words of data, a Merkle proof requires ten layers, or a total of 6380 gas.
+
+Looking for example at [Optimism](https://public-grafana.optimism.io/d/9hkhMxn7z/public-dashboard?orgId=1&refresh=5m), at writing L1 gas costs about 100 gwei and L2 gas costs 0.001 gwei (that is the normal price, it can rise with congestion). So for the cost of one L1 gas we can spend a hundred thousand gas on L2 processing. Assuming we don't overwrite storage, this means that we can write about five words to storage on L2 for the price of one L1 gas. For a single Merkle proof we can write the entire 1024 words to storage (assuming they can be calculated on chain to begin with, rather than provided in a transaction) and still have most of the gas left over.
```

Comment:
```suggestion
Looking for example at [Optimism](https://public-grafana.optimism.io/d/9hkhMxn7z/public-dashboard?orgId=1&refresh=5m), writing L1 gas costs about 100 gwei and L2 gas costs 0.001 gwei (that is the normal price, it can rise with congestion). So for the cost of one L1 gas we can spend a hundred thousand gas on L2 processing. Assuming we don't overwrite storage, this means that we can write about five words to storage on L2 for the price of one L1 gas. For a single Merkle proof we can write the entire 1024 words to storage (assuming they can be calculated on chain to begin with, rather than provided in a transaction) and still have most of the gas left over.
```